go/vt/vtctl/workflow/workflows.go

```diff
@@ -0,0 +1,672 @@
+		copyStates[i] = &vtctldatapb.Workflow_Stream_CopyState{
+			StreamId: streamId,
+			Table:    row["table_name"].ToString(),
+			LastPk:   row["lastpk"].ToString(),
+		}
+	}
+
+	return copyStates, nil
+}
+
+func (w *workflow) buildWorkflows(
+	ctx context.Context,
+	results map[*topo.TabletInfo]*tabletmanagerdatapb.ReadVReplicationWorkflowsResponse,
+	copyStatesByShardStreamId map[string][]*vtctldatapb.Workflow_Stream_CopyState,
+	req *vtctldatapb.GetWorkflowsRequest,
+) ([]*vtctldatapb.Workflow, error) {
+	workflowsMap := make(map[string]*vtctldatapb.Workflow, len(results))
+	workflowMetadataMap := make(map[string]*workflowMetadata, len(results))
+
+	for tablet, result := range results {
+		// In the old implementation, we knew we had at most one (0 <= N <= 1)
+		// workflow for each shard primary we queried. There might be multiple
+		// rows (streams) comprising that workflow, so we would aggregate the
+		// rows for a given primary into a single value ("the workflow",
+		// ReplicationStatusResult in the old types).
+		//
+		// In this version, we have many (N >= 0) workflows for each shard
+		// primary we queried, so we need to determine if each row corresponds
+		// to a workflow we're already aggregating, or if it's a workflow we
+		// haven't seen yet for that shard primary. We use the workflow name to
+		// dedupe for this.
+		for _, wfres := range result.Workflows {
+			workflowName := wfres.Workflow
+			workflow, ok := workflowsMap[workflowName]
+			if !ok {
+				workflow = &vtctldatapb.Workflow{
+					Name:         workflowName,
+					ShardStreams: map[string]*vtctldatapb.Workflow_ShardStream{},
+				}
+
+				workflowsMap[workflowName] = workflow
+				workflowMetadataMap[workflowName] = &workflowMetadata{
+					sourceShards: sets.New[string](),
+					targetShards: sets.New[string](),
+				}
+			}
+
+			metadata := workflowMetadataMap[workflowName]
+			err := w.scanWorkflow(ctx, workflow, wfres, tablet, metadata, copyStatesByShardStreamId, req.Keyspace)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	for name, workflow := range workflowsMap {
+		meta := workflowMetadataMap[name]
+		updateWorkflowWithMetadata(workflow, meta)
+
+		// Sort shard streams by stream_id ASC, to support an optimization
+		// in fetchStreamLogs below.
+		for _, shardStreams := range workflow.ShardStreams {
+			sort.Slice(shardStreams.Streams, func(i, j int) bool {
+				return shardStreams.Streams[i].Id < shardStreams.Streams[j].Id
+			})
+		}
+	}
+
+	if req.IncludeLogs {
+		var fetchLogsWG sync.WaitGroup
+
+		for _, workflow := range workflowsMap {
+			// Fetch logs for all streams associated with this workflow in the background.
+			fetchLogsWG.Add(1)
+			go func(ctx context.Context, workflow *vtctldatapb.Workflow) {
+				defer fetchLogsWG.Done()
+				w.fetchStreamLogs(ctx, req.Keyspace, workflow)
+			}(ctx, workflow)
+		}
+
+		// Wait for all the log fetchers to finish.
+		fetchLogsWG.Wait()
+	}
+
+	return maps.Values(workflowsMap), nil
```

Comment:
It might be worth sorting the slice here by keyspace/workflowName. Can be done later.